### PR TITLE
DF: update README

### PR DESCRIPTION
**Pull request by anastasiakaida**
This update is about the description of stage types.
The main reason to add this block is a declaration what stage in a dataflow
pipeline is responsible for data preparation before uploading (sink or
transform stage).

We consider to leave sink stages pretty simple. All data preparation
work should be done by pre-sink stage -- the last one transform stage in
any data pipeline.

### CODE REVIEW

**Comment by anastasiakaida** [comment]
@Evildoor during the previous meeting you told that you have something to discuss about this PR. Please, review. It would be nice to get your feedback. :)

File: Utils/Dataflow/README [hunk 1]
```diff
@@ -91,6 +91,30 @@ intended for making sure that information is consistent between ProdSys2 and
 ES. It gets a very basic set of metadata from ProdSys2, adds ES-related
 fields, and checks that it is present in ES rather than uploading it.
 
+================
+* Stages types *
```
Comment (Evildoor) [inline Utils/Dataflow/README]: I think that "stages" should be in singular form: "Stage types".

File: Utils/Dataflow/README [hunk 2]
```diff
@@ -91,6 +91,30 @@ intended for making sure that information is consistent between ProdSys2 and
...
+================
+There are three types of stages:
+
+1. Source stage - an initial stage in a dataflow pipeline that gets data
+from an external source (CDS, Rucio, AMI, etc.)
```
Comment (Evildoor) [inline Utils/Dataflow/README]: Missing period at the end of sentence: "... AMI, etc.).". I had my doubts about this, but it seems that this particular case have already been discussed: https://english.stackexchange.com/questions/23022/how-to-deal-with-abbreviations-like-etc-at-the-end-of-parentheses-which-are-c.

File: Utils/Dataflow/README [hunk 3]
```diff
@@ -91,6 +91,30 @@ intended for making sure that information is consistent between ProdSys2 and
...
+
+2. Transform stage - a stage that makes a set of different operations
```
Comment (Evildoor) [inline Utils/Dataflow/README]: 1) "... stage that PERFORMS a set of ..." or something else. "Stage makes operations" sounds unnatural.
2) I'm not sure that operations are always different. What if stage performs a single operation? What if it performs several, but all of them are very similar?

File: Utils/Dataflow/README [hunk 4]
```diff
@@ -91,6 +91,30 @@ intended for making sure that information is consistent between ProdSys2 and
...
+with metadata forming messages according to a particular pattern (e.g. select
+relevant metadata from a raw dataset, append metadata to a message, prepare
+messages to upload).
+
+2.1 Pre-sink stage - the last transform stage in a dataflow pipeline that
+prepares metadata to upload via sink stage; it awares about messages
```
Comment (Evildoor) [inline Utils/Dataflow/README]: "... via sink stage; it is aware of messages' formatting."

File: Utils/Dataflow/README (hunk 3 above)
Comment (mgolosova) [reply Utils/Dataflow/README]: "... -- these stages perform different operations..."?
The idea, I believe, was to emphasize that "transformation" actions can be really different for different cases. Or maybe not )

File: Utils/Dataflow/README [hunk 5]
```diff
@@ -91,6 +91,30 @@ intended for making sure that information is consistent between ProdSys2 and
...
+================
+* Stages types *
+================
+There are three types of stages:
+
+1. Source stage - an initial stage in a dataflow pipeline that gets data
+from an external source (CDS, Rucio, AMI, etc.)
+
+2. Transform stage - a stage that makes a set of different operations
+with metadata forming messages according to a particular pattern (e.g. select
```
Comment (mgolosova) [inline Utils/Dataflow/README]: ...on metadata...

File: Utils/Dataflow/README [hunk 6]
```diff
@@ -91,6 +91,30 @@ intended for making sure that information is consistent between ProdSys2 and
...
+relevant metadata from a raw dataset, append metadata to a message, prepare
+messages to upload).
+
+2.1 Pre-sink stage - the last transform stage in a dataflow pipeline that
+prepares metadata to upload via sink stage; it awares about messages
+formatting. All the messages preparations should be done by a pre-sink stage.
```
Comment (mgolosova) [inline Utils/Dataflow/README]: ...all message preparations...
(or "all the messages should be prepared by...", or "all the preparations")

But maybe "by the pre-sink stage" (or "by the pre-sink stages" -- to emphasize that there may be many pre-sinks), for the whole paragraph is a description of this specific type of stages and the last sentence refers to the same thing the reader is already aware about.

File: Utils/Dataflow/README [hunk 7]
```diff
@@ -91,6 +91,30 @@ intended for making sure that information is consistent between ProdSys2 and
...
+
+3. Sink stage - a stage that uploads data into the DKB storage (Virtuoso
+or ES). We consider that every single sink stage requires a particular
```
Comment (mgolosova) [inline Utils/Dataflow/README]: Better to keep the list open: "(Virtuoso, ES, ...)" -- for technically we are not limited to these two storages, and "sink stage" will remain "sink" even if it works with Neo4j or Cassandra ;)

"We consider" -> "we expect" / "it is supposed" / "every sink stage requires" ?

"particular" -> "specific"? Maybe it is just me, but "particular" sounds more like "something in question" for me, like "this particular case".

File: Utils/Dataflow/README [hunk 8]
```diff
@@ -91,6 +91,30 @@ intended for making sure that information is consistent between ProdSys2 and
...
+pre-sink stage that transform data and form ready-to-send messages with a
+particular structure. Sink stage only gets prepared messages and upload it
+without any re-formatting. It does not care about messages content.
+If message does not follow the template, it will not be uploaded.
```
Comment (mgolosova) [inline Utils/Dataflow/README]: "...most likely it will be rejected by the storage load API"?
If simply say "not uploaded", it is not clear who is responsible for "not uploading" the message.

File: Utils/Dataflow/README (hunk 7 above)
Comment (Evildoor) [inline Utils/Dataflow/README]: I suggest making this impersonal and changing some other things: "... or ES). Every single sink stage should be preceded by a corresponding pre-sink stage that ..."

File: Utils/Dataflow/README [hunk 9]
```diff
@@ -91,6 +91,30 @@ intended for making sure that information is consistent between ProdSys2 and
...
+relevant metadata from a raw dataset, append metadata to a message, prepare
+messages to upload).
+
+2.1 Pre-sink stage - the last transform stage in a dataflow pipeline that
+prepares metadata to upload via sink stage; it awares about messages
+formatting. All the messages preparations should be done by a pre-sink stage.
+
+3. Sink stage - a stage that uploads data into the DKB storage (Virtuoso
+or ES). We consider that every single sink stage requires a particular
+pre-sink stage that transform data and form ready-to-send messages with a
```
Comment (Evildoor) [inline Utils/Dataflow/README]: "transformS", "formS".

File: Utils/Dataflow/README [hunk 10]
```diff
@@ -91,6 +91,30 @@ intended for making sure that information is consistent between ProdSys2 and
...
+particular structure. Sink stage only gets prepared messages and upload it
```
Comment (Evildoor) [inline Utils/Dataflow/README]: "uploadS".

File: Utils/Dataflow/README (hunk 5 above)
Comment (Evildoor) [reply Utils/Dataflow/README]: This, in my opinion, is a part of the larger problem. Usually it is possible to say "perform operations with metadata", but in this case we have a construction "perform operations with metadata forming messages according to pattern", which can be interpreted in many ways (especially if a possibility of missing comma or two is considered), such as:
1. Perform operations, and in these operations metadata forms messages according to pattern.
2. Perform operations according to pattern, and in these operations metadata forms messages.
3. Perform operations with metadata and, by doing so, form messages according to pattern.
4. Perform operations with some of the metadata - the ones that form messages according to pattern.
I suggest rewording the sentence to avoid such ambiguity, and changing "with" to "on" looks like a good start, as it "cuts off" some "branches".

File: Utils/Dataflow/README (hunk 3 above)
Comment (Evildoor) [reply Utils/Dataflow/README]: I still doubt the word "different" here. Also, come to think of it - possibility (and existence of 016) of stage performing a single operation is an argument against "set of" as well, not only "different".
Emphasizing of possible difference - maybe it can be done elsewhere, or in another way?

Comment (anastasiakaida) [reply Utils/Dataflow/README]: @Evildoor As Marina said, transformation operations can be different, but I can emphasize it like this "...these stages perform a set of operations for different cases". Hope, it sounds correct.

File: Utils/Dataflow/README [hunk 11]
```diff
@@ -91,6 +91,30 @@ intended for making sure that information is consistent between ProdSys2 and
...
+* Stages types *
+================
+There are three types of stages:
+
+1. Source stage - an initial stage in a dataflow pipeline that gets data
+from an external source (CDS, Rucio, AMI, etc.)
+
+2. Transform stage - a stage that makes a set of different operations
+with metadata forming messages according to a particular pattern (e.g. select
+relevant metadata from a raw dataset, append metadata to a message, prepare
```
[the remark this answers is not in the document]
Comment (anastasiakaida) [reply Utils/Dataflow/README]: @mgolosova It means not already transformed, but uploaded with any source stage data. :)

**Comment by anastasiakaida** [comment]
@mgolosova, @Evildoor 
Now README is fixed according to your comments. Thank you!
Please, review. :)

File: Utils/Dataflow/README [hunk 12]
```diff
@@ -133,8 +133,34 @@ following requirements:
...
+1. Source stage - an initial stage in a dataflow pipeline that gets data
+from an external source (CDS, Rucio, AMI, etc.).
+
+2. Transform stage - these stages perform a set of operations for different
+cases on metadata and form output messages according to a particular pattern
+(e.g. select relevant metadata from a raw dataset, append metadata
+to a message, prepare messages for upload).
+
+2.1 Pre-sink stage - the last transform stage in a dataflow pipeline that
+prepares metadata for upload via sink stage; it awares about messages
```
Comment (Evildoor) [inline Utils/Dataflow/README]: "... ; is is aware of messages' formatting rules. ..."

File: Utils/Dataflow/README [hunk 13]
```diff
@@ -133,8 +133,34 @@ following requirements:
...
+==================
+There are three types of stages:
+
+1. Source stage - an initial stage in a dataflow pipeline that gets data
+from an external source (CDS, Rucio, AMI, etc.).
+
+2. Transform stage - these stages perform a set of operations for different
+cases on metadata and form output messages according to a particular pattern
+(e.g. select relevant metadata from a raw dataset, append metadata
+to a message, prepare messages for upload).
```
Comment (Evildoor) [inline Utils/Dataflow/README]: Maybe "... prepare a message for upload"? To follow the style of two previous examples.

File: Utils/Dataflow/README [hunk 14]
```diff
@@ -133,8 +133,34 @@ following requirements:
...
+==================
+* 3. Stage types *
+==================
+There are three types of stages:
+
+1. Source stage - an initial stage in a dataflow pipeline that gets data
+from an external source (CDS, Rucio, AMI, etc.).
+
+2. Transform stage - these stages perform a set of operations for different
+cases on metadata and form output messages according to a particular pattern
```
Comment (mgolosova) [inline Utils/Dataflow/README]: Maybe it worth to be completely reworded?
Something like
```
2. Transform stage -- transitional stage that performs single logical
operation on the metadata. Set (chain) of transform stages forms the whole
tarnsformation process for messages emitted by Source.
Transformation opeations may include: format change, derived values
calculation, message extention with metadata from additional source, etc.
```

File: Utils/Dataflow/README (hunk 13 above)
Comment (mgolosova) [reply Utils/Dataflow/README]: Maybe "preparation for upload" no to be mentioned here at all? I mean, we have "2.1. Pre-sink stage" section below -- it looks like a rightful place to talk about this for me...

File: Utils/Dataflow/README [hunk 15]
```diff
@@ -133,8 +133,34 @@ following requirements:
...
+(e.g. select relevant metadata from a raw dataset, append metadata
+to a message, prepare messages for upload).
+
+2.1 Pre-sink stage - the last transform stage in a dataflow pipeline that
```
Comment (mgolosova) [inline Utils/Dataflow/README]:
```suggestion
2.1. Pre-sink stage - the last transform stage in a dataflow pipeline that
```

File: Utils/Dataflow/README (hunk 3 above)
Comment (mgolosova) [reply Utils/Dataflow/README]: Resolving conversation in favor of the next iteration here: https://github.com/PanDAWMS/dkb/pull/278#discussion_r313291818

File: Utils/Dataflow/README (hunk 5 above)
Comment (mgolosova) [reply Utils/Dataflow/README]: New variant uses the `operations on metadata` version so I guess this discussion may be resolverd now.

File: Utils/Dataflow/README [hunk 16]
```diff
@@ -133,8 +133,34 @@ following requirements:
...
+* 3. Stage types *
+==================
+There are three types of stages:
+
+1. Source stage - an initial stage in a dataflow pipeline that gets data
+from an external source (CDS, Rucio, AMI, etc.).
+
+2. Transform stage - these stages perform a set of operations for different
+cases on metadata and form output messages according to a particular pattern
+(e.g. select relevant metadata from a raw dataset, append metadata
```
[the remark this answers is not in the document]
Comment (mgolosova) [reply Utils/Dataflow/README]: Sorry I missed related conversation above(https://github.com/PanDAWMS/dkb/pull/278#discussion_r304831028), so I resolve this conversation and will add a coment there.

File: Utils/Dataflow/README (hunk 11 above)
[the remark this answers is not in the document]
Comment (mgolosova) [reply Utils/Dataflow/README]: In terms of ATLAS-related dataflow the "raw dataset" (as well as any other "dataset") is more likely to be interpreted as the dataset itself (the RAW one, containing raw data from the detector, or any other). I guess these datasets do contain some metadata, butwe do not work with any datasets directly, so... I think it would be better to say not "dataset", but "dataset metadata" -- should sound less confusing.

File: Utils/Dataflow/README (hunk 12 above)
Comment (mgolosova) [reply Utils/Dataflow/README]: @Evildoor, seems like you missed my reply to the similar comment of yours above:
 (https://github.com/PanDAWMS/dkb/pull/278#discussion_r304834350). Could you take a look and let me (and @anastasiakaida) know what do you think?

Comment (Evildoor) [reply Utils/Dataflow/README]: I'd rather say that what's wrong here is the fact that I've made the comment and then (regardless of your reply's existence) forgot about it and made the same suggestion again. My bad.

File: Utils/Dataflow/README (hunk 4 above)
Comment (Evildoor) [reply Utils/Dataflow/README]: I think that "message formatting rules" is the best out of the three suggested options.

File: Utils/Dataflow/README (hunk 16 above)
[the remark this answers is not in the document]
Comment (anastasiakaida) [reply Utils/Dataflow/README]: @mgolosova And you asked (and I even answered ;) ), but there was a difference between definitions.
From my point of view, it was something like "not already transformed, but uploaded with any source stage data", I mean "data from an external source as it was", but I see -- in terms of ATLAS-related dataflow, it is a completely different thing. I suppose I can reword it to prevent misunderstanding.

**Comment by anastasiakaida** [comment]
@mgolosova, @Evildoor 
Fixed! Thank you! :)

File: Utils/Dataflow/README (hunk 12 above)
Comment (mgolosova) [reply Utils/Dataflow/README]: The conversation is not solved (I mean, the original comment of @Evildoor is still not addressed in the current version of the README).

File: Utils/Dataflow/README [hunk 17]
```diff
@@ -133,8 +133,35 @@ following requirements:
...
+* 3. Stage types *
+==================
+There are three types of stages:
+
+1. Source stage - an initial stage in a dataflow pipeline that gets data
+from an external source (CDS, Rucio, AMI, etc.).
+
+2. Transform stage - a transitional stage that performs single logical
+operation on the metadata. Set (chain) of transform stages forms the whole
+transformation process for messages emitted by Source. Transformation
```
Comment (Evildoor) [inline Utils/Dataflow/README]: "Emitted" is usually applied to energy/light/etc. How about "produced", "constructed", or something else?

**Comment by anastasiakaida** [comment]
> Few line comments to some particular moments, and one more general: maybe move the whole section to "REFERENCES"? It is said to contain "Conventions", and convention it is: what specific stages should look like to fit the architecture/concept. Somewhere after "streaming mode", I think?

Fixed. Now it is under "references".

File: Utils/Dataflow/README (hunk 1 above)
Comment (anastasiakaida) [reply Utils/Dataflow/README]: @Evildoor fixed

File: Utils/Dataflow/README (hunk 2 above)
Comment (anastasiakaida) [reply Utils/Dataflow/README]: According to the discussion:
>The dot in etc. If it's inside the parentheses, like in this case, then you should use two dots, since one is for the abbreviation, the other is for the sentence.

Fixed. Now it looks like you've described.

File: Utils/Dataflow/README (hunk 12 above)
Comment (anastasiakaida) [reply Utils/Dataflow/README]: @mgolosova My bad. Fixed now.

@Evildoor Thank you!

File: Utils/Dataflow/README (hunk 8 above)
Comment (anastasiakaida) [reply Utils/Dataflow/README]: @mgolosova 
Fixed! Thank you!

File: Utils/Dataflow/README (hunk 10 above)
Comment (anastasiakaida) [reply Utils/Dataflow/README]: @Evildoor 
Fixed!

File: Utils/Dataflow/README (hunk 17 above)
Comment (anastasiakaida) [reply Utils/Dataflow/README]: @Evildoor 
I think, "produced" may be correct enough for the context

**Comment by anastasiakaida** [comment]
@Evildoor @mgolosova 
I think, everything is fixed now. Please, check this PR (hope, according to @Evildoor comment, this will be the last one).
:)

File: Utils/Dataflow/README [hunk 18]
```diff
@@ -133,8 +133,35 @@ following requirements:
...
+* 3. Stage types *
+==================
+There are three types of stages:
+
+1. Source stage - an initial stage in a dataflow pipeline that gets data
+from an external source (CDS, Rucio, AMI, etc.).
+
+2. Transform stage - a transitional stage that performs single logical
+operation on the metadata. Set (chain) of transform stages forms the whole
+transformation process for messages produced by Source. Transformation
```
Comment (Evildoor) [inline Utils/Dataflow/README]: The word "Source" should not be capitalized: "source".

File: Utils/Dataflow/README [hunk 19]
```diff
@@ -133,8 +133,35 @@ following requirements:
...
+2.1. Pre-sink stage - the last transform stage in a dataflow pipeline that
+prepares metadata for upload via sink stage; it is aware of messages'
+formatting rules. All the preparations for upload should be done by
+the pre-sink stage.
+
+3. Sink stage - a stage that uploads data into the DKB storage
+(Virtuoso, ES, ...). Every single sink stage should be preceded by
+a corresponding pre-sink stage that transforms data and forms ready-to-send
+messages with a specific structure. Sink stage only gets prepared messages
+and uploads it without any re-formatting. It does not care about messages
```
Comment (mgolosova) [inline Utils/Dataflow/README]:
```suggestion
and uploads them without any re-formatting. It does not care about messages
```

File: Utils/Dataflow/README (hunk 18 above)
Comment (mgolosova) [reply Utils/Dataflow/README]: If it should go with lowercase initial letter, then I suggest to extend it to "source stage". Else one may take it as "metadata source" (just as in all the other cases where lowercase "source" appears), while what produces *messages* should clearly be the stage.

**Comment by mgolosova** [comment]
@anastasiakaida, last two comments left, I believe :) We almost there, thank you for bringing things to order!

File: Utils/Dataflow/README (hunk 18 above)
Comment (anastasiakaida) [reply Utils/Dataflow/README]: @mgolosova, @Evildoor
I suppose to change it into a "source stage". :)

**Comment by anastasiakaida** [comment]
@mgolosova Fixed! :)

**Comment by mgolosova** [comment]
@anastasiakaida, great! I have approved the PR, so now we wait for the final review from @Evildoor and if everything\`s fine it will be merged soon.
